pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -467,39 +450,36 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun) err
 	}
 	// Read the condition the way it was set by the Mark* helpers
 	after = pr.Status.GetCondition(apis.ConditionSucceeded)
-	pr.Status.TaskRuns = getTaskRunsStatus(pr, pipelineState)
+	pr.Status.TaskRuns = getTaskRunsStatus(pr, pipelineState.State())
 	logger.Infof("PipelineRun %s status is being set to %s", pr.Name, after)
 	return nil
 }
 
-// runNextSchedulableTask gets the next schedulable Tasks from the dag based on the current
-// pipeline run state, and starts them
-// after all DAG tasks are done, it's responsible for scheduling final tasks and start executing them
-func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.PipelineRun, d *dag.Graph, dfinally *dag.Graph, pipelineState resources.PipelineRunState, as artifacts.ArtifactStorageInterface) error {
-
+// runNextSchedulableTask gets the next schedulable Tasks based the pipeline run state and starts them
+func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.PipelineRun, pipelineState resources.ResolvedPipelineRun, as artifacts.ArtifactStorageInterface) error {
 	logger := logging.FromContext(ctx)
 	recorder := controller.GetEventRecorder(ctx)
 
-	var nextRprts []*resources.ResolvedPipelineRunTask
+	var (
+		nextRprts []*resources.ResolvedPipelineRunTask
+		err       error
+	)
 
 	// when pipeline run is stopping, do not schedule any new task and only
 	// wait for all running tasks to complete and report their status
-	if !pipelineState.IsStopping(d) {
-		// candidateTasks is initialized to DAG root nodes to start pipeline execution
-		// candidateTasks is derived based on successfully finished tasks and/or skipped tasks
-		candidateTasks, err := dag.GetSchedulable(d, pipelineState.SuccessfulOrSkippedDAGTasks(d)...)
+	if !pipelineState.IsStopping() {
```

Comment:
this check can be moved in the `GetNextTasks`

Reply:
I think this check is already redundant now, but I need to try it out, and I was planning to do so in a next PR, along with giving a single NextTasks method for the pipeline controller to use.

Reply:
sounds good, thanks 👏